adr/0026-schema.md

```diff
@@ -0,0 +1,569 @@
+# 25. Zarf Schema for 1
+
+Date: 2024-06-07
+
+## Status
+
+Proposed
+
+## Context
+
+Zarf currently does not have explicit schema versions. Any schema changes are embedded into Zarf and can change with new versions. There are several examples of deprecated keys throughout Zarf's lifespan such as:
+
+- `setVariable` deprecated in favor of `setVariables`
+- `scripts` deprecated in favor of `actions`
+- `required` soon to be deprecated in favor of `optional`.
+- `group` deprecated in favor of `flavor`, however these work functionally different and cannot be automatically migrated
+- `cosignKeyPath` deprecated in favor of specifying the key path at create time
+
+Zarf has not disabled any deprecated keys thus far. On create the user is always warned when using a deprecated field, however the field still exists in the schema and functions properly. Some of the deprecated keys can be migrated automatically as the old item is a subset or directly related to it's replacement. For example, setVariable is automatically migrated to a single item list of setVariables. This migration occurs on the zarf.yaml fed into the package during package create, however the original field is not deleted from the packaged zarf.yaml because the Zarf binary used in the airgap or delivery environment is not assumed to have the new schema fields that the deprecated key was migrated to.
+
+There are two problems this ADR aims to solve:
+- Supporting deprecated features forever is complex and time consuming
+- Dropping feature support can frustrate or confuse users
+
+## Decision
+
+Zarf will begin having proper schema versions. A top level key, `apiVersion`, will be introduced to allow users to specify the schema. At the release of v1 the only valid user input for `apiVersion` will be v1. Zarf will not allow users to build at the pre v1 version. `zarf package create` will fail if the user has deprecated keys or if `apiVersion` is missing and the user will be instructed to run the new `zarf dev update-schema` command. `zarf dev update-schema` will automatically migrate deprecated fields in the users `zarf.yaml` where possible. It will also add the apiVersion key and set it to v1.
```

Comment:
Why `apiVersion` instead of `schemaVersion` or just `version`?

Reply:
What decisions led to us supporting schema migration?

Reply:
apiVersion is used to keep similarity with the KRM.

https://github.com/kubernetes/design-proposals-archive/blob/main/architecture/resource-management.md

Reply:
Since we're going to have code in the backend anyway to transform deprecated fields into their new counterparts I feel it makes sense to expose that to the user to give them easier upgrade paths.

Additionally, Razzle already started on the code here https://github.com/defenseunicorns/zarf/pull/2211/files#diff-048b26f18d5c4dcd3484e355fb29fce16f91b49940754f2d5096f7c9494994c3.